punchcut/src/scaler.rs

```diff
@@ -0,0 +1,172 @@
+use super::{source::*, Context, Error, Hinting, NormalizedCoord, PathSink, Result, Variation};
+
+use read_fonts::{
+    types::{Fixed, GlyphId, Tag},
+    TableProvider,
+};
+
+use core::{borrow::Borrow, str::FromStr};
+
+/// Builder for configuring a glyph scaler.
+pub struct ScalerBuilder<'a> {
+    context: &'a mut Context,
+    font_id: Option<u64>,
+    size: f32,
+    hint: Option<Hinting>,
+}
+
+impl<'a> ScalerBuilder<'a> {
+    /// Creates a new builder for configuring a scaler with the given context.
+    pub fn new(context: &'a mut Context) -> Self {
+        context.coords.clear();
+        context.variations.clear();
+        Self {
+            context,
+            font_id: None,
+            size: 0.0,
+            hint: None,
+        }
+    }
+
+    /// Sets a unique font identifier for hint state caching.
+    pub fn font_id(mut self, font_id: Option<u64>) -> Self {
+        self.font_id = font_id;
+        self
+    }
+
+    /// Sets the font size in pixels per em units.
+    ///
+    /// A size of 0.0 will disable scaling and result in glyphs defined in font units.
+    pub fn size(mut self, size: f32) -> Self {
+        self.size = size.abs();
+        self
+    }
+
+    /// Sets the hinting mode.
+    pub fn hint(mut self, hint: Option<Hinting>) -> Self {
+        self.hint = hint;
+        self
+    }
+
+    /// Specifies a variation with a set of normalized coordinates.
+    ///
+    /// This will clear any variations specified with the variations method.
+    pub fn coords<I>(self, coords: I) -> Self
+    where
+        I: IntoIterator,
+        I::Item: Borrow<NormalizedCoord>,
+    {
+        self.context.variations.clear();
+        self.context.coords.clear();
+        self.context
+            .coords
+            .extend(coords.into_iter().map(|v| *v.borrow()));
+        self
+    }
+
+    /// Adds the sequence of variation settings. This will clear any variations
+    /// specified as normalized coordinates.
+    pub fn variations<I>(self, variations: I) -> Self
+    where
+        I: IntoIterator,
+        I::Item: Into<Variation>,
+    {
+        self.context.coords.clear();
+        self.context
+            .variations
+            .extend(variations.into_iter().map(|v| v.into()));
+        self
+    }
+
+    /// Builds a scaler using the currently configured settings
+    /// for the specified font.
+    pub fn build(self, font: &impl TableProvider<'a>) -> Scaler<'a> {
+        if !self.context.variations.is_empty() {
```

Comment:
Easier to read with less nesting; suggest a helper fn that early-exits with guard clauses for some of the nop cases such as no variation settings, no fvar, or no axes.

Reply:
I'll break it up a bit.

Reply:
Thanks for doing the work for me. Done :)